### PR TITLE
Add functions.MaybeAll

### PR DESCRIPTION
**Pull request by kalman**
This is like functions.All but returns an error if any function returns
an error.

### CODE REVIEW

File: go/util/functions/all.go [hunk 1]
```diff
@@ -6,16 +6,40 @@ package functions
...
-// Runs all functions in |fs| in parallel, and returns when all functions have returned.
+// All runs all functions in parallel, and returns when all functions have
+// returned.
 func All(fs ...func()) {
 	wg := &sync.WaitGroup{}
 	wg.Add(len(fs))
-	for _, f_ := range fs {
-		f := f_
+	for _, f := range fs {
+		f := f
```
Comment (kalman) [inline go/util/functions/all.go]: golint complained about the `f_`. I believe `f := f` is idiomatic anyway.

File: go/util/functions/all.go [hunk 2]
```diff
@@ -6,16 +6,40 @@ package functions
...
 		go func() {
+			defer wg.Done()
```
Comment (kalman) [inline go/util/functions/all.go]: Oops

**Comment by ehalpern** [comment]
Jenkins: test this

File: go/util/functions/all.go [hunk 3]
```diff
@@ -6,16 +6,40 @@ package functions
...
 			f()
-			wg.Done()
 		}()
 	}
 	wg.Wait()
 }
+
+// MaybeAll runs all functions in parallel, and returns when all functions have
+// returned. If any function returns an error, returns that error, but not
```
Comment (aboodman) [inline go/util/functions/all.go]: This seems like a weird contract ... it waits until all functions have returned, but only returns one of the errors at random? It seems like you'd either want (a) it returns when the first one errors, or (b) it waits until all are done and returns all the errors.

File: go/util/functions/all.go [hunk 4]
```diff
@@ -6,16 +6,40 @@ package functions
...
+// until all functions have returned.
+func MaybeAll(fs ...func() error) (err error) {
+	wg := &sync.WaitGroup{}
+	wg.Add(len(fs))
+	errMtx := sync.Mutex{}
+	for _, f := range fs {
+		f := f
+		go func() {
+			defer wg.Done()
+			fErr := f()
```
Comment (aboodman) [inline go/util/functions/all.go]: Do you want to catch panics?

**Pull request closed without merging**